src/aap_eda/services/project/scm.py

```diff
@@ -0,0 +1,248 @@
+        final_url = url
+        secret = ""
+        key_file = None
+        if credential:
+            inputs = inputs_from_store(credential.inputs.get_secret_value())
+            secret = inputs.get("password", "")
+            key_data = inputs.get("ssh_key_data", "")
+
+            final_url = cls.build_url(
+                url,
+                inputs.get("username", ""),
+                secret,
+                key_data,
+            )
+
+            if key_data:  # ssh
+                key_file = tempfile.NamedTemporaryFile("w+t")
+                key_file.write(key_data)
+                key_file.flush()
+                vars.append(f"key_file={key_file.name}")
+                key_password = inputs.get("ssh_key_unlock")
+                if key_password:
+                    cls.decrypt_key_file(key_file.name, key_password)
+
+        if not verify_ssl:
+            _executor.ENVIRON["GIT_SSL_NO_VERIFY"] = "true"
+
+        vars.append(f"scm_url={final_url}")
+        if branch:
+            vars.append(f"scm_branch={branch}")
+        if refspec:
+            vars.append(f"scm_refspec={refspec}")
+        if depth:
+            vars.append(f"depth={depth}")
+
+        vars_str = " ".join(vars)
+
+        cmd = ["-e", f"'{vars_str}'", "-t", "update_git"]
+        logger.info("Cloning repository: %s", url)
+        try:
+            git_hash = _executor(cmd)
+        except ScmError as e:
+            msg = str(e)
+            if secret:
+                msg = msg.replace(secret, "****", 1)
+            logger.warning("SCM clone failed: %s", msg)
+            raise ScmError(msg) from None
+        finally:
+            if key_file:
+                key_file.close()
+        instance = cls(path, _executor=_executor)
+        instance.git_hash = git_hash
+        return instance
+
+    def _execute_cmd(self, cmd: Iterable[str]):
+        return self._executor(cmd, cwd=self.root)
+
+    @classmethod
+    def build_url(
+        cls, url: str, user: str, password: str, ssh_key: str
+    ) -> str:
+        result = urlparse(url)
+        domain = result.netloc.split("@")[-1]
+        scheme = result.scheme
+        path = result.path
+
+        if ssh_key and scheme == "" and path.startswith("git@"):
+            return url
+        elif user and password:
+            domain = f"{user}:{password}@{domain}"
+
+        unparsed = (
+            scheme,
+            domain,
+            path,
+            result.params,
+            result.query,
+            result.fragment,
+        )
+        return urlunparse(unparsed)
+
+    @classmethod
+    def decrypt_key_file(cls, key_file: str, password: str) -> str:
+        cmd = f'{KEYGEN_COMMAND} -p -P {password} -N "" -f {key_file}'
+        child = pexpect.spawn(cmd)
+        index = child.expect(["Failed to load key", pexpect.EOF])
+        if index == 0:
+            raise ScmError("Incorrect passhprase for the private key")
+
+
+class PlaybookExecutor:
```

Comment:
This class doesn't seem to be reusable for playbooks like its name suggests because we are expecting interactive outputs strongly related with a specific operation/playbook. Should be renamed to avoid misunderstandings?